### ISSUE TITLE
Handle clipped email body

### ISSUE THREAD

**Issue by robin-thomas** [KartikTalwar/gmail.js#564]
Clipped emails show on the UI like `[Message clipped]  View entire message`

When trying to use `gmail.new.get.email_data(id)` on such an email, it returns an empty `content_html`.

Shouldn't the `content_html` mention that the email body is clipped?

How to handle this case (against an email without any content)?

**Comment by josteink** [comment]
> Clipped emails show on the UI like [Message clipped] View entire message

If so, I’m not sure when or how it is fetched. 

> Shouldn't the content_html mention that the email body is clipped?

I agree some flag might be useful here. 

> How to handle this case (against an email without any content)?

Use the view_email event and inspect the provided DOM.Email instance?

Optionally fetch and parse email_source() but I suspect that will involve much more work.

**Comment by robin-thomas** [comment]
> Use the view_email event and inspect the provided DOM.Email instance?

I'm using the below code to read the email data.
```
gmail.observe.on('open_email', async function(id) {
  let data = gmail.new.get.email_data(id);
}
```

Shouldn't the email body be already in the DOM by then?

**Comment by josteink** [comment]
Use view_email instead. Param IS the DomEmail for that mail and it should have email body available.

**Comment by robin-thomas** [comment]
Another thing I noticed is, if there are emojis in the email, google loads the email again (with `goomoji`), which means the `email_id` changes after the load.

Is there any way to detect whether that's going to happen?

**Comment by robin-thomas** [comment]
When I used `view_email` and opening the email the first time, results in `undefined` id.

![](https://i.ibb.co/5nKdrxC/gmailjs.png)

It works if I open the email once again.

**Comment by josteink** [comment]
But why do you need the ID? The instance should contain the body itself without a need for any ID. 

Can you show your full code?

**Comment by josteink** [comment]
Also: See docs here:

https://github.com/KartikTalwar/gmail.js/blob/master/README.md#gmaildomemailemail_el-or-email_id

**Comment by robin-thomas** [comment]
I do not need the ID, but I noticed that it's `undefined`. I'm trying to get the `subject` and `content_html` of the email.

```
  gmail.observe.on('view_email', async function(domEmail) {
      const data = domEmail.data();
      const label = data.subject;   <-- its failing here, saying data is "undefined"
      const content = data.content_html;
});
```

**Comment by josteink** [comment]
You already have the data in the DomEmail.  Read the docs I linked. 

Don’t bother with data(). It’s deorecated anyway.

**Comment by josteink** [comment]
Oh right. Subject *is* missing, because it belongs to the thread. 

In which case you may need to just stall fetching (setTimeout) and retry later against gmail.new.get.email_data(DomEmail).

**Comment by robin-thomas** [comment]
Can you give a sample code for the `setTimeout`?

I would like to know when to stop calling `gmail.new.get.email_data(DomEmail)` (if in case, its not loaded the first time).

>You already have the data in the DomEmail. Read the docs I linked.
I was looking for a way to get the `subject`. Yes.

**Comment by josteink** [comment]
I’m on the phone, busy, nowhere near a computer so getting a full code sample is asking for more than I can provide right now. 

I had just stall for 10-100 ms to give the GUI thread time to complete rendering. 

If that doesn’t work, you will probably have to create a promise-like construct to repeat for some retries and then call back when it finally had found a value. 

I plan to add something like that for the 1.0 release, but it’s not ready quite yet.

**Comment by robin-thomas** [comment]
> In which case you may need to just stall fetching (setTimeout) and retry later against gmail.new.get.email_data(domEmail).

Its `gmail.new.get.email_data(email_id)`.
https://github.com/KartikTalwar/gmail.js#gmailnewgetemail_dataemail_id
But I don't have the id as its `undefined`

Maybe I misunderstood something.

**Comment by josteink** [comment]
No. It’s not just the ID. The DomEmail will do fine. 

 I recently (yesterday or so) updated the code to allow more ways to identify an email, to provide a more ergonomic and reliable API. 

The docs remain to be fully updated though.

**Comment by robin-thomas** [comment]
Ah, I'm using version `0.9.4` of `gmail-js`. Seems like this is the latest version available on npm.
In this version, if I pass `domEmail` as input, I get `undefined` as output.

**Comment by josteink** [comment]
If you for now use git master, 0.9.5 will probably be out in a few days and have dysfunctionality ready to use.

**Comment by robin-thomas** [comment]
So once `0.9.5` is out, all I have to do is to perform the below code?

```
gmail.observe.on('view_email', async function(domEmail) {
  while (gmail.new.get.email_data(domEmail) === undefined) {
    await sleep(100); // wait for 100 ms.
  }

  const subject = domEmail.dom('subject');
  const body = domEmail.body();
});
```

Also, is `body` the same as `content_html`?

**Comment by josteink** [comment]
No. Obtain subject from the EmailData-instance. 

Also, I’m not sure if you the async constructions blocks the main thread or not, so in cases like that I prefer using settimeout to ensure that there’s clean separation between the event handler and the awaiting code.

**Comment by josteink** [comment]
I pushed 0.9.5 now.

See the boilerplate project for something which works (at least for me) using that version.

**Comment by robin-thomas** [comment]
I modified per your suggestions. Can you kindly check again?

```
gmail.observe.on('view_email', async function(domEmail) {
      let emailData = null;
      const intervalId = setInterval(() => {
        emailData = gmail.new.get.email_data(domEmail);
        if (emailData !== null && emailData !== undefined) {
          clearInterval(intervalId);
        }
      }, 100);  // call every 100ms until we get the email

      const label = emailData.subject;
      const content = emailData.content_html;
}
```

**Comment by josteink** [comment]
That’s overly complicated and if you check the  boilerplate project you will see what do you want to do, done in two lines of code.

**Comment by robin-thomas** [comment]
> That’s overly complicated and if you check the boilerplate project you will see what do you want to do, done in two lines of code.

@josteink
You mean here: https://github.com/wolfch/gmailjs-node-boilerplate/blob/selectize_input_demo/src/extension.js ?

Also, when I try the below code, 
```
gmail.observe.before('send_message', function(url, body, data, xhr) {

});
```
Both `body` and `data` and `xhr.xhrParams.body_params` are in weird form (meaning the keys in the object are numbers).

![](https://i.ibb.co/XYZPqnz/gmail-js-sendemail.png)

**Comment by josteink** [comment]
Use observe.after, not before. It should contain parsed data. 

Observe.before send_email sadly won’t work like before, since new Gmail was launched. 

Looking at your example, I’m not sure what is wrong, but I can’t look at it until later. 

Try some basic debugging? Step into the last function succeeding and see where it fails?

**Comment by robin-thomas** [comment]
@josteink I tried the `after` and when I checked the `response` object, I got the below json object.

![](https://i.ibb.co/r2KMnRD/gmailjs-resp.png)

Why is not human readable?

**Comment by josteink** [comment]
After Google changed all the underlying data-protocols for "New Gmail" we had to do quite a bit of rewriting just to keep things working mostly as expected.

Some events (and *especially* `"send message"`) are affected in ways we cannot do much about.

Basically a sent email is not sent... Because the user can still cancel it. It's not a *real* email which is sent for real until later. So we can't get email-data on that. And we don't have an event for when it becomes a real email.

This means that at this stage you will just have to dig the JSON for the message-id (`msg-a-r:175834758934`) and keep polling gmail.new.get.email_data(msgId) until you get something, and sometimes maybe even navigate to the email itself to get it into the DOM (to force it into the GmailJS cache).

In a extension for my company I've wrapped this up in a series of promises which makes it rather clean, but none of that code is part of GmailJS as of now.

On this turf... You're on your own, and better make due with what you get. 

I agree the ergonomics of this is subpar though. So if you want to improve any part of these aspects, feel free to investigate the codebase, try to make it better and send in some PRs!
